### PR TITLE
Global config withdrawal fee

### PR DESCRIPTION
**Pull request by ebatsell**

### CODE REVIEW

**Review by coachchucksol — CHANGES_REQUESTED** [review]
All looks good, we're also going to want the following:

[] - `set_fee_wallet` ix
[] - `fee_admin` field in config that has the authority to update the fee as well as the fee wallet - this is so we can give this to a specific committee ( Double check with @buffalu )
[] - Any lingering tests/lints

File: vault_sdk/src/instruction.rs [hunk 1]
```diff
@@ -201,6 +205,20 @@ pub enum VaultInstruction {
...
+    #[account(0, writable, name = "config")]
+    #[account(1, signer, name = "admin")]
+    SetProgramFee {
+        new_fee_bps: u16
+    },
+
+    /// Sets the program fee wallet for the vault program
+    #[account(0, writable, name = "config")]
+    #[account(1, signer, name = "config_fee_admin")]
+    SetConfigFeeWallet {
```
Comment (buffalu) [inline vault_sdk/src/instruction.rs:218]: missing sdk/tests?

File: vault_program/src/set_program_fee.rs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    account_info::AccountInfo, entrypoint::ProgramResult, msg, program_error::ProgramError,
+    pubkey::Pubkey,
+};
+
+/// Sets the program fee for the vault program.
+///
+/// Specification:
+/// - The fee can only be changed by the config admin. The config admin must sign the transaction.
+/// - The transaction shall fail if the new fee exceeds MAX_FEE_BPS.
+/// - The fee change must be acceptable according to the Config::check_fee_change_ok method.
```
Comment (buffalu) [inline vault_program/src/set_program_fee.rs:15]: its Vault:: check_fee_change_ok right now

File: integration_tests/tests/vault/burn_withdrawal_ticket.rs [hunk 3]
```diff
@@ -553,6 +589,7 @@ mod tests {
                 &vault_staker_withdrawal_ticket,
                 &get_associated_token_address(&vault_staker_withdrawal_ticket, &vault.vrt_mint),
                 &get_associated_token_address(&vault.fee_wallet, &vault.vrt_mint),
+                &get_associated_token_address(&config.program_fee_wallet, &vault.vrt_mint),
                 MIN_AMOUNT_OUT,
             )
             .await;
```
Comment (buffalu) [inline integration_tests/tests/vault/burn_withdrawal_ticket.rs:595]: can you add test here + in burn for handling the config withdrawal fee?

File: vault_core/src/config.rs [hunk 4]
```diff
@@ -69,8 +86,11 @@ impl Config {
             deposit_withdrawal_fee_cap_bps: PodU16::from(Self::DEFAULT_FEES_CAP_BPS),
             fee_rate_of_change_bps: PodU16::from(Self::DEFAULT_FEE_RATE_OF_CHANGE_BPS),
             fee_bump_bps: PodU16::from(Self::DEFAULT_FEE_BUMP_BPS),
+            program_fee_bps: PodU16::from(program_fee_bps),
+            program_fee_wallet,
+            fee_admin: admin, // TODO: Change this to a different admin
```
Comment (buffalu) [inline vault_core/src/config.rs:91]: can remove todo, i think fine for now

File: vault_core/src/vault.rs [hunk 5]
```diff
@@ -794,9 +797,23 @@ impl Vault {
             return Err(VaultError::VaultInsufficientFunds);
         }
 
-        let fee_amount = self.calculate_withdraw_fee(amount_in)?;
+        let program_fee_amount = Config::calculate_program_fee(program_fee_bps, amount_in)?;
```
Comment (buffalu) [inline vault_core/src/vault.rs:800]: what're thoughts on calculating the vault fee as a % of amount_in - program_fee_amount? i think makes the code slightly simpler? but some tradeoffs. lmk wyt

File: vault_core/src/vault.rs [hunk 6]
```diff
@@ -794,9 +797,23 @@ impl Vault {
...
+        let mut vault_fee_amount = self.calculate_withdraw_fee(amount_in)?;
+
+        // Prioritize program fee over vault fee if they exceed the amount in
+        if program_fee_amount
```
Comment (buffalu) [inline vault_core/src/vault.rs:804]: this seems wrong?

File: vault_core/src/vault.rs [hunk 7]
```diff
@@ -2004,6 +2028,6 @@ mod tests {
     #[test]
     fn test_burn_with_fee_zero_amount() {
         let mut vault = make_test_vault(0, 0, 1000, 1000, DelegationState::default());
-        assert_eq!(vault.burn_with_fee(0, 0), Err(VaultError::VaultBurnZero));
+        assert_eq!(vault.burn_with_fee(0, 0, 0), Err(VaultError::VaultBurnZero));
```
Comment (buffalu) [inline vault_core/src/vault.rs:2031]: i don't see any tests with first parameter > 0? need to add that

Comment (ebatsell) [reply vault_core/src/vault.rs:2079]: added

File: vault_sdk/src/instruction.rs (hunk 1 above)
Comment (ebatsell) [reply vault_sdk/src/instruction.rs:218]: added

File: vault_core/src/vault.rs (hunk 6 above)
Comment (ebatsell) [reply vault_core/src/vault.rs:805]: thought we had decided that we will ensure program fee gets its full share, and then vault fee will get whatever is left if they sum above 100 - is that not desired? or is the implementation here wrong?

File: vault_core/src/vault.rs (hunk 5 above)
Comment (ebatsell) [reply vault_core/src/vault.rs:801]: what would calculate_vault_fee take in and return in this case?

File: vault_core/src/vault.rs (hunk 6 above)
Comment (buffalu) [reply vault_core/src/vault.rs:805]: actually this can just be simplified to: let fee = min(amount_in, program_fee_amount + vault_fee_amount)?

Comment (ebatsell) [reply vault_core/src/vault.rs:805]: this gets us the total fee amount but we'd still need logic to adjust vault fee if needed right?

File: vault_program/src/set_program_fee.rs [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        msg!("Config admin does not match");
+        return Err(VaultError::VaultConfigAdminInvalid.into());
+    }
+
+    if new_fee_bps > MAX_FEE_BPS {
+        msg!("New fee exceeds maximum allowed fee");
+        return Err(ProgramError::InvalidInstructionData);
+    }
+
+    Vault::check_fee_change_ok(
```
Comment (buffalu) [inline vault_program/src/set_program_fee.rs:40]: i think okay to remove this, wdyt? will likely be controlled by DAO

File: vault_core/src/vault.rs [hunk 9]
```diff
@@ -858,13 +879,24 @@ impl Vault {
             .and_then(|x| x.checked_add(self.vrt_ready_to_claim_amount()))
             .ok_or(VaultError::VaultOverflow)?;
 
-        let fee_amount = self.calculate_withdraw_fee(vrt_reserve)?;
+        let vault_fee_amount = self.calculate_withdraw_fee(vrt_reserve)?;
```
Comment (buffalu) [inline vault_core/src/vault.rs:882]: need to remind myself why we need to do the fee calculation here...

**Review by buffalu — COMMENTED** [review]
looking closer

**Review by buffalu — APPROVED** [review]

File: vault_program/src/set_program_fee.rs (hunk 8 above)
Comment (coachchucksol) [reply vault_program/src/set_program_fee.rs:40]: I think I agree - if the DAO controls it, it's okay to get rid of the restriction

**Review by coachchucksol — CHANGES_REQUESTED** [review]

File: vault_program/src/burn_withdrawal_ticket.rs [hunk 10]
```diff
@@ -26,23 +26,29 @@ pub fn process_burn_withdrawal_ticket(
...
-    let (required_accounts, optional_accounts) = accounts.split_at(11);
-    let [config, vault_info, vault_token_account, vrt_mint, staker, staker_token_account, vault_staker_withdrawal_ticket_info, vault_staker_withdrawal_ticket_token_account, vault_fee_token_account, token_program, system_program] =
+    let (required_accounts, optional_accounts) = accounts.split_at(12);
+    let [config, vault_info, vault_token_account, vrt_mint, staker, staker_token_account, vault_staker_withdrawal_ticket_info, vault_staker_withdrawal_ticket_token_account, vault_fee_token_account, program_fee_token_account, token_program, system_program] =
         required_accounts
     else {
         return Err(ProgramError::NotEnoughAccountKeys);
     };
 
+    msg!("loading config");
```
[the remark this answers is not in the document]
Comment (ebatsell) [reply vault_program/src/burn_withdrawal_ticket.rs:36]: 😳😳😳

**Review by coachchucksol — APPROVED** [review]
lgtm

**Pull request merged**